renormalizer/cv/spectra_cv.py

```diff
@@ -9,130 +9,181 @@
+    cores: number of cores to be used in multiprocessing calculation
+    obj: SpectraZtCV or SpectraFtCV
+    """
+    logger.info(f"{len(freq_reg)} total frequency points to do")
+    spectra = []
 
-def unwrap_self(arg, **kwarg):
-    return SpectraCv.cv_solve(*arg, **kwarg)
+    if cores > 1:
+        # multiprocessing
+        if importlib.util.find_spec("cupy"):
+            multiprocessing.set_start_method('forkserver', force=True)
+        pool = Pool(processes=cores)
+        logger.info(f"{cores} multiprocess parallelization activated")
+        for i_spec in pool.imap(obj.cv_solve, freq_reg):
+            spectra.append(i_spec)
+
+    elif cores == 1:
+        # single process
+        for omega in freq_reg:
+            spectra.append(obj.cv_solve(omega))
+            #obj.cv_mps.ensure_left_canon()
+            obj.clear_res()
+    else:
+        assert False
+
+    return spectra
 
 
 class SpectraCv(object):
     def __init__(
         self,
         mol_list,
         spectratype,
-        freq_reg,
         m_max,
         eta,
-        method="1site",
-        procedure_cv=None,
-        cores=1
+        h_mpo = None,  
+        method = "1site",
+        procedure_cv = None,
+        rtol = 1e-5,
+        b_mps = None,
+        e0 = None,
+        cv_mps = None,
     ):
-        self.freq_reg = freq_reg
-        self.m_max = m_max
-        self.eta = eta
+        
         self.mol_list = mol_list
-        assert spectratype in ["abs", "emi"]
+        
+        assert spectratype in ["abs", "emi", None]
         self.spectratype = spectratype
-        # percent used to update correction vector for each sweep process
+        
+        self.m_max = m_max
+        self.eta = eta
+        
+        # Hamiltonian
+        if h_mpo is None:
+            self.h_mpo = Mpo(mol_list)
+        else:
+            self.h_mpo = h_mpo
+
+        assert method in ["1site", "2site"]
+        self.method = method
+        logger.info(f"cv optimize method: {method}")
+        
+        # percent used to update correction vector for each isweep process
         # see function mps.lib.select_basis
         if procedure_cv is None:
             procedure_cv = [0.4, 0.4, 0.2, 0.2, 0.1, 0.1] + [0] * 45
         self.procedure_cv = procedure_cv
-        self.method = method
-        self.cores = cores
-    '''
-    def check_qn(self, X):
-        check_op = Mpo.onsite(self.mol_list, r"a^\dagger a", dipole=False)
-        check_1 = X.conj().dot(check_op.apply(X)) / X.conj().dot(X)
-        print('Quantum number of X', check_1)
-    '''
-    def run(self):
-        if importlib.util.find_spec("cupy"):
-            multiprocessing.set_start_method('forkserver', force=True)
-        start = time.time()
-        pool = Pool(processes=self.cores)
-        logger.info(f"{len(self.freq_reg)} total frequency to do")
-        logger.info(f"{self.cores} multiprocess parallelization activated")
-        freq_reg = self.freq_reg
-        spectra = []
```

Comment:
Just to mention. Previously, I did so because I thought that we have to copy many objects of the class "SpectraCv" to make sure that each "object" exist independently and the variables (for example, self.cv_mps) will not be shared by different multiprocesses at the same time, which brings conflicts. However, as you implemented here @jjren, we will not meet such circumstance.  Although I am pretty sure that I did meet troubles at that time by using the logic as yours and I am not able to reproduce the problems any more. Thanks for the improvement.

Reply:
To me, the `multiprocessing` is actually a little bit tricky. I'm really not a big fan of it.
This a demo to show that the instance variable would be changed in a child process if `chunksize >1`.  However, If `chunksize =1`, it's ok (imap has default `chunksize=1`).  The same problem has been proposed in https://stackoverflow.com/questions/48332391/instance-methods-with-multiprocessing-pool  
But there is no good answer. It seems that each chunk will have an independent instance because of `pickle`. 
The code works now. But I think maybe we should replace it with `mpi4py` someday because MPI is more sophisticated from my point of view.
```
from multiprocessing import Pool
class c:
    def __init__(self):
        self.a = []

    def print_a(self,i):
        self.a.append(i)
        print(self.a,i, self)
        return i
pool = Pool(processes=2)
hello = c()
for j in pool.imap(hello.print_a, list(range(10)),chunksize=2):
    print(j)
pool.close()
pool.join()
```
output
[0] 0 <__main__.c object at 0x2b0c5bed0d30>
[2] 2 <__main__.c object at 0x2b0c5bed0d30>
[0, 1] 1 <__main__.c object at 0x2b0c5bed0d30>
[2, 3] 3 <__main__.c object at 0x2b0c5bed0d30>
[4] 4 <__main__.c object at 0x2b0c5a76f588>
[6] 6 <__main__.c object at 0x2b0c5a76f588>
[6, 7] 7 <__main__.c object at 0x2b0c5a76f588>
[4, 5] 5 <__main__.c object at 0x2b0c5a76f588>
[8] 8 <__main__.c object at 0x2b0c5a76f748>
[8, 9] 9 <__main__.c object at 0x2b0c5a76f748>

if chunksize=1
[0] 0 <__main__.c object at 0x2b0c5bed0588>
[1] 1 <__main__.c object at 0x2b0c5bed0588>
[2] 2 <__main__.c object at 0x2b0c5a76f748>
[3] 3 <__main__.c object at 0x2b0c5adfe6d8>
[4] 4 <__main__.c object at 0x2b0c5a76f748>
[5] 5 <__main__.c object at 0x2b0c5adfe6d8>
[6] 6 <__main__.c object at 0x2b0c5a76f470>
[7] 7 <__main__.c object at 0x2b0c5a76f588>
[8] 8 <__main__.c object at 0x2b0c5a4d4f28>
[9] 9 <__main__.c object at 0x2b0c5a76f588>

Reply:
What we've implemented (use instance method in `map` or `imap`) is actually not a standard way to use multiprocessing in its document (not yet documented at least ).
I guess mpi4py could easily handle it.

Reply:
understood.